domain/application/state/application.go

```diff
@@ -2193,6 +2194,412 @@ FROM v_revision_updater_application_unit
+			}
+
+			updates = append(updates, setApplicationConfig{
+				ApplicationUUID: ident.ID.String(),
+				Key:             k,
+				Value:           config[k].Value,
+				TypeID:          typeID,
+			})
+
+		}
+		var inserts []setApplicationConfig
+		for k, v := range config {
+			if _, ok := currentM[k]; ok {
+				continue
+			}
+
+			typeID, err := encodeConfigType(v.Type)
+			if err != nil {
+				return internalerrors.Errorf("encoding config type: %w", err)
+			}
+
+			inserts = append(inserts, setApplicationConfig{
+				ApplicationUUID: ident.ID.String(),
+				Key:             k,
+				Value:           v.Value,
+				TypeID:          typeID,
+			})
+		}
+
+		// We have to check the foreign key constraint on each request, as
+		// each one is optional, bar the last query.
+
+		if len(removals) > 0 {
+			if err := tx.Query(ctx, deleteStmt, removals, ident).Run(); internaldatabase.IsErrConstraintForeignKey(err) {
+				return applicationerrors.ApplicationNotFound
+			} else if err != nil {
+				return internalerrors.Errorf("deleting config: %w", err)
+			}
+		}
+		if len(inserts) > 0 {
+			if err := tx.Query(ctx, insertStmt, inserts).Run(); internaldatabase.IsErrConstraintForeignKey(err) {
+				return applicationerrors.ApplicationNotFound
+			} else if err != nil {
+				return internalerrors.Errorf("inserting config: %w", err)
+			}
+		}
+		for _, update := range updates {
+			if err := tx.Query(ctx, updateStmt, update).Run(); internaldatabase.IsErrConstraintForeignKey(err) {
+				return applicationerrors.ApplicationNotFound
+			} else if err != nil {
+				return internalerrors.Errorf("updating config: %w", err)
+			}
+		}
+
+		if err := tx.Query(ctx, settingsStmt, setApplicationSettings{
+			ApplicationUUID: ident.ID.String(),
+			Trust:           trust,
+		}).Run(); internaldatabase.IsErrConstraintForeignKey(err) {
+			return applicationerrors.ApplicationNotFound
+		} else if err != nil {
+			return internalerrors.Errorf("updating settings: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return internalerrors.Errorf("setting application config: %w", err)
+	}
+	return nil
+}
+
+// UnsetApplicationConfigKeys removes the specified keys from the application
+// config. If the key does not exist, it is ignored.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+func (st *State) UnsetApplicationConfigKeys(ctx context.Context, appID coreapplication.ID, keys []string) error {
+	db, err := st.DB()
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	ident := applicationID{ID: appID}
+
+	// This isn't ideal, as we could request this in one query, but we need to
+	// perform multiple queries to get the data. First is to get the application
+	// availability, second to just get the application overlay config for the
+	// charm config and the application settings for the trust config.
+	appQuery := `
+SELECT &applicationID.*
+FROM application
+WHERE uuid = $applicationID.uuid;
```

Comment:
Same here

Reply:
Why is this one missed?

Are we able to unset application config keys for dead applications?